acquia_lift.context.inc

```diff
@@ -0,0 +1,156 @@
+<?php
+
+/**
+ * @file acquia_lift.context.inc
+ * Provides functions needed for handling visitor contexts.
+ */
+
+/**
+ * Gets all options for mappable contexts
+ *
+ * @return array
+ *   An array of options keyed by the context type
+ */
+function _acquia_lift_get_context_options() {
+  $contexts = array(
+    'taxonomy_context' => _acquia_lift_get_taxonomy_context_options(),
+  );
+  return array_filter($contexts);
+}
+
+/**
+ * Gets all the callbacks for finding the mapped value of a context type.
+ * The callback receives the the node and the mapped value.
+ *
+ * @return array
+ *   An array of callables.
+ */
+function _acquia_lift_get_mapping_callbacks() {
+  return array(
+    'taxonomy_context' => '_acquia_lift_get_mapped_taxonomy_context'
+  );
+}
+
+/**
+ * Gets all of the node fields that include taxonomy terms.
+ *
+ * @return array
+ *   An array of contextual options that can be used for grouped output.
+ */
+function _acquia_lift_get_taxonomy_context_options() {
+  if (!module_exists('taxonomy')) {
+    return array();
+  }
+
+  $vocabularies = taxonomy_get_vocabularies();
+  $options = array();
+  foreach ($vocabularies as $vocabulary) {
+    $machine_name = $vocabulary->machine_name;
+    $options[$machine_name] = array(
+      'name' => $vocabulary->name,
+      'machine_name' => $vocabulary->machine_name,
+      'group' => 'Taxonomy',
+      'id' => $vocabulary->vid,
+    );
+  }
+  return $options;
+}
+
+/**
+ * Handles the page context request data.
+ *
+ * @param $node
+ *   (optional) If passed then the taxonomy terms for the selected node
+ *   are added to the page taxonomy context.
+ * @param bool $primary
+ *   True if this is the primary node displayed for the page, false otherwise.
+ * @return array
+ *   Returns the taxonomy context for the page.
+ */
+function _acquia_lift_get_page_context($node = NULL, $primary = FALSE) {
+  $page_context = &drupal_static(__FUNCTION__);
+  if (!isset($page_context)) {
+    $page_context = array();
+  }
+
+  // Add the content type of the page if this is the primary node displayed.
+  if ($primary && empty($page_context['post_id'])) {
+    $page_context['content_type'] = $node->type;
+    $page_context['content_title'] = $node->title;
+    $page_context['published_date'] = $node->created;
+    $page_context['post_id'] = $node->nid;
+    $account = user_load($node->uid);
+    $page_context['author'] = $account->name;
+    $page_context['page_type'] = 'node page';
+    $page_context['type'] = $node->type;
```

Comment:
It was taken from this code here: https://github.com/acquia/visitor_actions/blob/7.x-1.x/visitor_actions.module#L672